pkg/provisioner/ironic/raid.go

```diff
@@ -138,7 +144,57 @@ func buildTargetSoftwareRAIDCfg(volumes []metal3v1alpha1.SoftwareRAIDVolume) (lo
 }
 
 // BuildRAIDCleanSteps build the clean steps for RAID configuration from BaremetalHost spec
-func BuildRAIDCleanSteps(raid *metal3v1alpha1.RAIDConfig) (cleanSteps []nodes.CleanStep) {
+func BuildRAIDCleanSteps(raidInterface string, target *metal3v1alpha1.RAIDConfig, existed *metal3v1alpha1.RAIDConfig) (cleanSteps []nodes.CleanStep, err error) {
+	err = checkRAIDConfigure(raidInterface, target)
+	if err != nil {
+		return nil, err
+	}
+
+	// No RAID
+	if raidInterface == "no-raid" {
+		return
+	}
+
+	// Software RAID
+	if raidInterface == "agent" {
+		cleanSteps = append(
+			cleanSteps,
+			[]nodes.CleanStep{
+				{
+					Interface: "raid",
+					Step:      "delete_configuration",
+				},
+				{
+					Interface: "deploy",
+					Step:      "erase_devices_metadata",
+				},
+			}...,
+		)
```

Comment:
Looks like the above steps are configured always, in case of the software RAID, regardless the current status vs the target one (whereas for the hardware RAID, later, a `DeepEqual` check is performed to detect an eventual discrepancy). Wouldn't be required a similar check also here before starting to assemble the cleansteps?

Reply:
Because I think software RAID may be modified in the process of use, and we can't guarantee that it is consistent with the content in the status.

Reply:
Software RAID may be changed by the OS. We cannot guarantee that the actual SoftwareRAID and status are the same, so I will always delete SoftwareRAID and clear `Status.Provisioning.RAID.SoftwareRAIDVolumes`.

Reply:
This happens only during the `Preparing` state, so I guess it would be anyway fine for now - still it would be nice to have in future a more robust way to apply software RAID only when required.

Reply:
Done.